Add TailwindCSS colors (and script to update them)

jmuchovej opened the pull request:
👋 Following up on #24, I've added the colors from TailwindCSS (an incredibly popular CSS framework).

Their code is MIT licensed and their colors don't appear to have a different license, so it seems that these colors are also MIT licensed.

I've also run the tests using `julia --project=. tests/runtests.jl` and received the following output:
```
Test Summary: | Pass  Total  Time
tailwind      |    2      2  0.1s
Test Summary: | Pass  Total  Time
xkcd          |    2      2  0.0s
Test Summary: | Pass  Total  Time
Resene        |    2      2  0.1s
Test Summary: | Pass  Total  Time
NBS           |    2      2  0.0s
Test Summary: | Pass  Total  Time
X11           |    2      2  0.0s
Test Summary: | Pass  Total  Time
CSS3          |    2      2  0.0s
Test Summary: | Pass  Total  Time
Crayola       |    2      2  0.0s
Test Summary: | Pass  Total  Time
WinsorNewton  |    2      2  0.0s
Test Summary: | Pass  Total  Time
named lookup  |    4      4  0.1s
Test Summary:                        | Pass  Total  Time
colorant_str macro gets named colors |    4      4  0.0s
Test Summary:                                  | Pass  Total  Time
colorant_str macro gets original functionality |    2      2  0.0s
```

jmuchovej committed:
Adding TailwindCSS colors

jmuchovej committed:
Adding TailwindCSS colors

jmuchovej committed:
Add JS script to extract TailwindCSS colors

Adds JavaScript script to track the colors that ship with the `tailwindcss`
`npm` package. Also edits `.gitignore` to prevent tracking `node_modules`
and `package-lock.json`

jmuchovej committed:
Merge branch 'add-tailwindcss' of github.com:jmuchovej/NamedColors.jl into add-tailwindcss

kellertuer commented on test/loading.jl:
```diff
@@ -1,4 +1,16 @@
 
+@testset "tailwind" begin
```
Is this a copy paste error? It seems to check twice the same here, but the second still has the xkcd name from the next test?

jmuchovej replied:
Yes, it definitely was. 😅

kellertuer commented on scripts/exportTailwindColors.cjs:
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env node
```
Just curious: What is a `.cjs` file?
This seems to be a JS file and how does it have to be run? Ccould that be documented?

In my personal favour, an import that is Julia-based would even be cooler, if possible. Since this seems to not to a HTTP call or so, maybe the loading of coolers has to be done in JS though?

jmuchovej replied:
> What is a `.cjs` file?

`.cjs` files are functionally JavaScript files, _but_ convey that imports are done via `CommonJS` (which is what `node` understands by default).

> ... how does it have to be run? Could that be documented?

👀 I thought I did this. 😅 (I see that I in-fact did not.) It could definitely be documented. Though let's circle-back after this detour down using Julia x NodeJS.

I’d also love to use a Julia script to do this. 😅

> Since this seems to not to a HTTP call or so, maybe the loading of coolers has to be done in JS though?

So, the file that contains this color data is [available here][tailwind-colors]. Though it's mostly JSON-like, there are a few bits that would break every JSON parser I'm familiar with. (I'd also prefer to avoid doing some RegEx-like matching since `colors` has reliably exported by Tailwind for the last two major versions.)

However, I see now that [`NodeJS.jl`][nodejs.jl] exists, so it's possible I could reimplement the update script in Julia by-proxy? Unless, you were thinking of a pure Julia implementation (I'm not sure how feasible this would be, sadly.)

[nodejs.jl]: https://github.com/davidanthoff/NodeJS.jl
[tailwind-colors]: https://github.com/tailwindlabs/tailwindcss/blob/master/src/public/colors.js

kellertuer replied:
With “A file in julia” I was thinking of a Julia script that (as independent as possible) updates the colors with something like `julia update-tailwind.jl` – not that you should implement that completely yourself, but with `NodeJS.jl` for example.

jmuchovej replied:
Ah. I didn't give `NodeJS.jl` a proper look. So, all it appears to do is provide a wrapper for Julia to run `node` and `npm` commands, so the JS code seems it would still need to exist. I think the alternative would be to roll a brittle solution that is liable to break if the folks at Tailwind ever change the path/format of `colors`.

kellertuer replied:
I would also be fine with a thorough `Readme` how to update the colors, so this can be done, without reaching out to you ;) I thought a Julia script might be the easiest.

jmuchovej replied:
Sorry – I didn't see your comment about `julia update-tailwind.jl` before posting mine. 😅 

I'm happy to keep the JS script but add a Julia script alongside it that actually does the updates (via `NodeJS.jl`). 👍

kellertuer replied:
I am open for any solution, as long as a non-JS person as me has an easy way to run the update :)

jmuchovej replied:
Something I just considered (since Tailwind recently updated the colors). The colors are subject to change – I'm not sure if other colors in `NamedColors` are liable to change (often). (I've been using Tailwind for the last 4 years and this is the first time I recall it changing, so it might not be a very frequent thing.)

So, this leads me to: NodeJS has a way to install specific versions (via the `package.json`) or we can install the latest Tailwind colors. **Is there a preferred route for `NamedColors`?**

Personally, I'm okay with the potentially moving target (installing the latest release of `tailwindcss`), but I dunno if there's some existing thoughts in the context of `NamedColors` about this.

kellertuer replied:
Since it would be a manual update triggered by a user, I think it is reasonable to take the latest release. If that changes, the user that has a reason to manually update them should also do a PR with a new version (a breaking one).

kellertuer reviewed (commented):
Overall this looks good to me; I noticed that the sources in the Readme are not up to date (even missing my last Paul Tol colours update *ehem*), could you at least add your source there as well?
(Feel free to list https://personal.sron.nl/~pault/ as well, since I forgot that).

jmuchovej committed:
Remove accidental copy-paste of `xkcd` test set

jmuchovej committed:
Adding links to TailwindCSS colors (and Paul Tol's)

jmuchovej committed:
Remove lingering `package.json`. Remove from `.gitignore`

jmuchovej committed:
Update `load_tailwind` to support comments

jmuchovej committed:
Remove file I/O from tailwind export JS

jmuchovej committed:
Add `update-tailwind.jl` for auto-TailwindCSS updating

jmuchovej commented:
So, I've added the link to Tailwind's (and Paul Tol's) colors in the README. I also removed the copy-paste error in `test/loading.jl`. Hold-off on accepting any changes 'cause Tailwind changed the colors and I forgot to verify that the tests pass.

I was able to extract most of the JS code into a Julia script – so if that's preferable, let me know and I'll remove the JS code and opt for the Julia script.

(Otherwise, I can trim the Julia code back down to the ~10 line script it was and let JS do its thing.)

To get the Julia script running, I needed to add `JSON` and `NodeJS`, but I'm not sure how Julia packaging works with "dev dependencies" (i.e., `JSON` and `NodeJS` aren't needed to use `NamedColors`, so I didn't wanna muddy the `Project.toml` if unnecessary).

jmuchovej committed:
Fix Tailwind tests (since there are 242 colors now)

kellertuer commented:
If the update script is in its own folder, it could get its own Project toml for example? Just within the `scripts/` folder do `Pkg.activate(".")` (or `] activate .` and add the packages there. If you want to be very precise add even a `[compat]` section to that project toml as well.

Sure it should not be additional dependencies for NamedColors.

But the rest already looks very good :)

jmuchovej committed:
Add deps for Tailwind update script

jmuchovej commented:
Done! Should I leave the JS code for reference, or take it out entirely? (I left it as reference mostly in case I ever needed to go back, but I could also make it a gist and just link to it from the `update-tailwind.jl` file?)

If you're good with leaving the JS code as reference, then this is ready to merge. 🎉

kellertuer commented:
If the Julia script now does the Job well, we could remove the old JS code to avoid confusion? So maybe a gist and linking it might be indeed the less confusing resolution here.

Besides that – I agree, we are nearly done – thanks for reacting to all my ideas and comments.

jmuchovej committed:
Remove JS Tailwind update, link to Gist.

Removes the JS script used to fetch Tailwind colors from their JS package and adds a link to a private GitHub Gist in the source of `scripts/update-tailwind.jl`.

jmuchovej commented:
Removed! 🙂 I linked the Gist in the header of `scripts/update-tailwind.jl`. At worst, someone would have to have to dig through the git history to find the original JS code (if they needed it). ¯\_(ツ)_/¯

kellertuer reviewed (approved):
🚀
I will wait a bit if some other developers have something to comment, but to me it looks good and I would merge it maybe end of the week?

We could also set the patch version one up (in the main project.toml) to register this as a new patch version.

kellertuer commented:
Could you maybe as a last step increase the patch number in the project toml? Them I would merge it and register a new version :)

jmuchovej committed:
Bump patch version

Closes #24.

kellertuer commented:
Thanks 👍
